bpo-33462: Add __reversed__ to dict and dict views

remilapeyre opened the pull request:
[bpo-33462](https://www.bugs.python.org/issue33462) - Add `__reverse__` to `dict`

See https://bugs.python.org/issue33462

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-33462](https://www.bugs.python.org/issue33462) -->
https://bugs.python.org/issue33462
<!-- /issue-number -->

Commit:
bpo-33462 - Add __reverse__ to dict

Commit:
Fix tests

Commit:
Add __reversed__ to dict views

vstinner commented on Objects/dictobject.c:
```diff
@@ -3651,6 +3652,113 @@ PyTypeObject PyDictIterItem_Type = {
+    PyDictKeysObject *k;
+    PyDictObject *d = di->di_dict;
+
+    if (d == NULL)
```
The PEP 7 now requires { ... } for if blocks.

vstinner commented on Objects/dictobject.c:
```diff
@@ -3651,6 +3652,113 @@ PyTypeObject PyDictIterItem_Type = {
+    sizeof(dictiterobject),
+    0,
+    /* methods */
+    (destructor)dictiter_dealloc,               /* tp_dealloc */
```
For new types, I now suggest to use the ".tp_dealloc = ...," syntax.

Commit:
Fix PEP7 issues

Commit:
Add Misc/NEWS.d entry

selik commented on Lib/test/test_dict.py:
```diff
@@ -1222,6 +1278,12 @@ def iter_and_mutate():
         self.assertRaises(RuntimeError, iter_and_mutate)
 
+    def test_reversed(self):
+        d = {i: i for i in range(20)}
```
Because integers have a special `hash()` behavior, perhaps a better test would be to use a few letters: "abcdefg". This shouldn't matter for the current dict implementation, but might make the test more robust against future implementation changes.

```
def test_reversed(self):
    keys = list('abcdefg')
    d = dict.fromkeys(keys)
    r = reversed(d)
    self.assertEqual(list(r), keys[::-1])
    self.assertRaises(StopIteration, next, r)
```

selik commented on Lib/test/test_enumerate.py:
```diff
@@ -160,9 +160,9 @@ def __getitem__(self, i):
                 return 5
-        for data in 'abc', range(5), tuple(enumerate('abc')), A(), range(1,17,5):
+        for data in ('abc', range(5), tuple(enumerate('abc')), A(),
+                    range(1,17,5), {i: i for i in range(20)}):
```
Again, `dict.fromkeys('abcdefg')` might be better here than `{i: i for i in range(20)}`.

Historically, dicts using int keys ended up sorted by accident. It's hard to imagine this being a problem, but it might give a false positive if the dict implementation changes later.

Commit:
Use non integer keys in tests

Commit:
Add reversible dict iterators to _collections_abc

methane commented on Objects/dictobject.c:
```diff
@@ -3327,9 +3328,25 @@ dictiter_new(PyDictObject *dict, PyTypeObject *itertype)
+            di->di_pos = dict->ma_used - 1;
+        }
+        else if (dict->ma_keys->dk_nentries) {
+            di->di_pos = dict->ma_keys->dk_nentries - 1;
```
why?  When dict->ma_used is zero, no need to scan entries.

methane commented on PC/python3.def:
```diff
@@ -87,6 +87,9 @@ EXPORTS
   PyDictIterKey_Type=python38.PyDictIterKey_Type DATA
+  PyDictRevIterKey_Type=python38.PyDictRevIterKey_Type DATA
+  PyDictRevIterItem_Type=python38.PyDictRevIterItem_Type DATA
+  PyDictRevIterValue_Type=python38.PyDictRevIterValue_Type DATA
```
Should these values be exposed by DLL?

remilapeyre replied:
Hi, if my memory is good (so I may absolutly be wrong) tests were not passing on windows before I added the types in this file. PyListRevIter_Type is here too so I thought those three new types may need to be here too but I'm not sure what puporse this file serves and if it is really necessary.

methane commented on Objects/dictobject.c:
```diff
@@ -3697,6 +3924,32 @@ dictiter_reduce(dictiterobject *di, PyObject *Py_UNUSED(ignored))
+    "dict_reverseitemiterator",
+    sizeof(dictiterobject),
+    .tp_dealloc = (destructor)dictiter_dealloc,
+    .tp_getattro = PyObject_GenericGetAttr,
```
Can't we skip defining `.tp_getattro`?

remilapeyre replied:
Hi @methane, thanks for the comments. I'm fixing the other issues. For this attribute, I'm not sure to understand precisely how `tp_getattro` is used and set it here because it was set on `PyDictKeys_Type`. As I understand it, it's the implementation of `getattr`. Is that right?

methane replied:
When NULL is set to members, PyType_Ready fills them to default values.
And `PyObject_GenericGetAttr` is the default value.

When we skip specifying all defaults explicitly, programmers can focus on
meaningful part of the code.

remilapeyre replied:
Thank you so much for the clarification, I read `PyType_Ready` definition and it all makes sense now.

The documentation at https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_getattro for this field says 

>An optional pointer to the get-attribute function.
> 
> The signature is the same as for PyObject_GetAttr(). It is usually convenient to set this field to PyObject_GenericGetAttr(), which implements the normal way of looking for object attributes.

Would it makes sense to propose a change from *"It is usually convenient to set this field to PyObject_GenericGetAttr(), which implements the normal way of looking for object attributes."* to something like *"If not set, `PyType_Ready` will set this field to PyObject_GenericGetAttr(), which implements the normal way of looking for object attributes."* for this field and the ones alike?

methane commented on Objects/dictobject.c:
```diff
@@ -3327,9 +3328,25 @@ dictiter_new(PyDictObject *dict, PyTypeObject *itertype)
+        itertype == &PyDictRevIterItem_Type ||
+        itertype == &PyDictRevIterValue_Type) {
+        if (dict->ma_used) {
+            di->di_pos = dict->ma_used - 1;
```
This seem wrong.

```
d = {'a':1, 'b':2, 'c':3}
del d['b']
```

ma_used is 2, and dk_nentries is 3.  dk_nentries-1 should be used.

methane commented on Objects/dictobject.c:
```diff
@@ -3651,6 +3668,210 @@ PyTypeObject PyDictIterItem_Type = {
+    PyObject *key;
+    Py_ssize_t i;
+    PyDictKeysObject *k;
+    PyDictObject *d = di->di_dict;
```
Since we use C11 from Python 3.6, I prefer [DeclareVariablesAtFirstUse](http://wiki.c2.com/?DeclareVariablesAtFirstUse) style for new code.

Commit:
Fix starting position

Commit:
Remove exposed types

methane commented on Objects/dictnotes.txt:
```diff
@@ -86,9 +86,9 @@ While making a dictionary more sparse reduces collisions, it impairs
-when it is created and once by dealloc().
+__iter__(), __reversed__(), iterkeys(), iteritems(), itervalues(),
+and update(). Also, every dictionary iterates at least twice, once for
+the memset() when it is created and once by dealloc().
```
Oh, this note was wrong.  Please don't touch.  I will rewrite this paragraph.

methane commented on Lib/_collections_abc.py:
```diff
@@ -305,6 +311,11 @@ def __subclasshook__(cls, C):
+Reversible.register(dict_keys)
+Reversible.register(dict_values)
+Reversible.register(dict_items)
+
```
What is `isinstance({}, Reversible)` output without register?

Since Reversible uses `__subclasshook__`, explicit registering will be redundant.

methane commented on Objects/dictobject.c:
```diff
@@ -3651,6 +3667,205 @@ PyTypeObject PyDictIterItem_Type = {
+}
+
+static PyObject *
+dictreviter_iternextvalue(dictiterobject *di)
```
Three `dictreviter_iternextXXX` have most same code.
Since reviter is very rarely used compared to normal iter, please optimize for maintenance cost than execution speed.

remilapeyre replied:
Indeed thanks, I removed the useless functions.

Commit:
Remove useless registion ofdict from Reversible classes in collections.abc

Commit:
Refactor dictreviter_iternextXXXX functions

Commit:
Revert uneeded notes modifications

methane committed:
revert _abc_collections

methane commented on Objects/dictobject.c:
```diff
@@ -3670,10 +3800,16 @@ dictiter_reduce(dictiterobject *di, PyObject *Py_UNUSED(ignored))
+        else if (Py_TYPE(di) == &PyDictRevIterValue_Type)
+            element = dictreviter_iternext(&tmp);
         else
             Py_UNREACHABLE();
```
Can we use switch-case here?

remilapeyre replied:
I'm not sure it's possible don't `case`s needs to be int consts?

I tried and got this output from the compiler: `error: expression is not an integer constant expression`.

methane replied:
You're right.

methane commented on Objects/dictobject.c:
```diff
@@ -3322,22 +3323,37 @@ dictiter_new(PyDictObject *dict, PyTypeObject *itertype)
+    }
+    else {
+        di->di_pos = 0;
+    }
```
This if statements looks ugly.

```
di->di_pos = 0;

if ((itertype == &PyDictRevIterKey_Type ||
     itertype == &PyDictRevIterItem_Type ||
     itertype == & PyDictRevIterValue_Type) && di->ma_used > 0)
{
    di->di_pos = dict->ma_keys->dk_nentries - 1;
}
```

This looks better, IMO.

This is matter of taste.  If you prefer current code, no need to rewrite.

remilapeyre replied:
This is indeed far better, thanks for noticing that

methane reviewed (commented):
You need to update document and what's new too.

Commit:
Simplify branching logic in dictiter_new

Commit:
Document reverse dict iterators

Commit:
Merge remote-tracking branch 'origin/master'

merwok commented on Doc/library/stdtypes.rst:
```diff
@@ -4237,6 +4237,11 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
+   .. describe:: reversed(d)
+
+      Return an reversed iterator over the keys of the dictionnary. This is a
+      shortcut for``iter(d.keys())``.
```
A shortcut for ``reversed(d.keys())`` you mean?

remilapeyre replied:
Indeed, I will fix this.

I saw that many tests are not passing anymore, I must have made something wrong during the merge. I will add a patch.

remilapeyre replied:
@merwok Is the documentation good enough once this mistake will be fixed?

merwok commented on Doc/library/stdtypes.rst:
```diff
@@ -4262,28 +4267,40 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
 
-   Dictionaries preserve insertion order.  Note that updating a key does not
-   affect the order.  Keys added after deletion are inserted at the end. ::
+   Dict preserves insertion order and are reversible.  Note that updating key
```
I think I would keep this text as it was before, and add a new short block to say “Dictionaries are reversible” and show the code examples.

Commit:
Call PyType_Ready on reversed dict iterator types

Commit:
Fix typo in WhatsNew

Commit:
Fix typo in documentation

Commit:
Update documentation

methane commented on Doc/library/stdtypes.rst:
```diff
@@ -4237,6 +4237,11 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
+
+      Return an reversed iterator over the keys of the dictionnary. This is a
+      shortcut for ``reversed(d.keys())``.
+
```
This section should be right after `describe:: iter(d)`.
``versionadded::`` is also helpful.

remilapeyre replied:
Looking at other sections of the documentation, I thought it was supposed to be ordered alphabetically, isn't that so?

methane commented on Lib/test/test_dict.py:
```diff
@@ -1071,6 +1071,62 @@ def test_valuesiterator_pickling(self):
+            it = reversed(data)
+            d = pickle.dumps(it, proto)
+            it = pickle.loads(d)
+            self.assertEqual(sorted(it), sorted(data))
```
Why `sorted`?  Can't we use `list`?

remilapeyre replied:
I look at the other tests like `test_valuesiterator_pickling` to write my own, they were certainly written before dict were ordered.

It would indeed be probably better to test with `list`, after all, we do respect order now.
I will change the tests I've written. Do you want me to change the other tests as well or prefer not to?

methane replied:
If you can.  Not mandatory.

remilapeyre reviewed (commented):
That's not a problem, after all I can live with `reversed(iter(d.values()))`. Thanks for your help so far.

methane commented on Doc/library/stdtypes.rst:
```diff
@@ -4237,6 +4237,11 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
+   .. describe:: reversed(d)
+
+      Return an reversed iterator over the keys of the dictionnary. This is a
+      shortcut for ``reversed(d.keys())``.
```
`an`? `dictionnary`?

methane commented on Doc/library/stdtypes.rst:
```diff
@@ -4365,6 +4399,10 @@ An example of dictionary view usage::
+   >>> list(reversed(keys))
+   ['spam', 'bacon']
+   >>> list(reversed(values))
+   [500, 1]
```
No need for adding examples here.

Commit:
Remove sorted usage from dict's tests

Dictionaries are officialy sorted since Python 3.7 and sorted usage
in tests can now be removed as the order is deterministic.

Commit:
Fix typo in dict documentation

Commit:
Remove dict.__reversed__ examples

methane commented on Doc/library/stdtypes.rst:
```diff
@@ -4284,6 +4289,27 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
+      [('four', 4), ('three', 3), ('two', 2), ('one', 42)]
+
+   .. versionchanged:: 3.8
+      Dictionaries and dictionary views are now reversible.
```
This versionchanged tag seems duplicated.  Remove `and dictionary views`.

methane commented on Doc/library/stdtypes.rst:
```diff
@@ -4284,6 +4289,27 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
+      >>> d["one"] = 42
+      >>> d
+      {'one': 42, 'two': 2, 'three': 3, 'four': 4}
+      >>> list(reversed(d.values()))
```
`d["one"] = 42` is example for "update value for a key doesn't change insertion order of a key".
It is not related reversed.  Remove these line and keep example minimal.
People don't read long example.

methane commented on Objects/dictobject.c:
```diff
@@ -3658,6 +3669,120 @@ PyTypeObject PyDictIterItem_Type = {
+            }
+            PyTuple_SET_ITEM(result, 0, key); /* steals reference */
+            PyTuple_SET_ITEM(result, 1, value); /* steals reference */
+
```
Remove this empty line

methane commented on Objects/dictobject.c:
```diff
@@ -3658,6 +3669,120 @@ PyTypeObject PyDictIterItem_Type = {
+
+        }
+        return result;
+    } else {
```
```
}
else {
```

Commit:
Fix coding style

Commit:
Clean documentation

Commit:
Merge remote-tracking branch 'origin/master'
